src/library/scala/collection/BitSetLike.scala

```diff
@@ -104,35 +114,91 @@ trait BitSetLike[+This <: BitSetLike[This] with SortedSet[Int]] extends SortedSe
 
   def iterator: Iterator[Int] = iteratorFrom(0)
 
-  override def keysIteratorFrom(start: Int) = new AbstractIterator[Int] {
-    private var current = start
-    private val end = nwords * WordLength
-    def hasNext: Boolean = {
-      while (current != end && !self.contains(current)) current += 1
-      current != end
+  override def keysIteratorFrom(start: Int) = {
+    val s = Math.max(0, start)
+
+    new AbstractIterator[Int] {
+      var wordIndex = s >> LogWL
+      val wordCount = nwords
+      var thisWord = if (wordIndex < wordCount)
+                       word(wordIndex) & (-1L << s)
+                     else 0L
+
+      def hasNext: Boolean = {
+        while (thisWord == 0 && wordIndex < wordCount) {
+          wordIndex += 1
+          if (wordIndex < wordCount)
+            thisWord = word(wordIndex)
+        }
+        wordIndex < wordCount
+      }
+      def next(): Int =
+        if (hasNext) {
+          val indexInWord = jLong.numberOfTrailingZeros(thisWord)
+          thisWord ^= (1L << indexInWord)
+          wordIndex * WordLength + indexInWord
+        } else Iterator.empty.next()
     }
-    def next(): Int =
-      if (hasNext) { val r = current; current += 1; r }
-      else Iterator.empty.next()
   }
 
   override def foreach[U](f: Int => U) {
-    /* NOTE: while loops are significantly faster as of 2.11 and
-       one major use case of bitsets is performance. Also, there
-       is nothing to do when all bits are clear, so use that as
-       the inner loop condition. */
-    var i = 0
-    while (i < nwords) {
-      var w = word(i)
-      var j = i * WordLength
-      while (w != 0L) {
-        if ((w&1L) == 1L) f(j)
-        w = w >>> 1
-        j += 1
+    var wordIndex = 0
+    val wordCount = this.nwords
+    while (wordIndex < wordCount) {
+      var thisWord = word(wordIndex)
+      val baseIndex = wordIndex * WordLength
+      while (thisWord != 0L) {
+        val indexInWord = jLong.numberOfTrailingZeros(thisWord)
+        thisWord ^= (1L << indexInWord)
+        f(baseIndex + indexInWord)
```

Comment:
Have you considered a cycle with less data dependency:
```
// Adopted from: https://lemire.me/blog/2019/05/03/really-fast-bitset-decoding-for-average-densities/
while (thisWord != 0L) {
  f(jLong.numberOfTrailingZeros(thisWord) + baseIndex)
  thisWord &= thisWord - 1
}
```

Reply:
I will try that in a benchmark